File: typescript-playwright/src/api/CommentApi.ts
import type { APIRequestContext, APIResponse } from "@playwright/test";
import { DEFAULT_ANSWER_TIME_MS } from "../utils/constants";
import type { CaptchaData, CommentPayload } from "../utils/types";
import { API_URL } from "../utils/constants";
import { mergeHeaders } from "../utils/headers";

export class CommentApi {
  private readonly request: APIRequestContext

  constructor(request: APIRequestContext) {
    this.request = request;
  }

  async getCaptcha(): Promise<APIResponse> {
    return this.request.get(`${API_URL}/api/comments/captcha`);
  }

  async getComments(photoId: number, page = 0, size = 5): Promise<APIResponse> {
    return this.request.get(`${API_URL}/api/comments`, {
      params: { photoId, page, size },
    });
  }

  async create(
    payload: CommentPayload,
    captcha: CaptchaData,
    customHeaders?: Record<string, string>,
  ): Promise<APIResponse> {
    const answerTimeMs = captcha.answerTimeMs ?? DEFAULT_ANSWER_TIME_MS;

    return this.request.post(`${API_URL}/api/comments`, {
      data: payload,
      headers: {
        ...mergeHeaders(),
        "X-Captcha-Session-Id": captcha.sessionId,
        "X-Captcha-Answer": String(captcha.answer),
        "X-Answer-Time-Ms": String(answerTimeMs),
        ...customHeaders,
      },
    });
  }
}

File: python-pytest/src/api/comment_api.py
from __future__ import annotations

import requests

from src.utils.constants import API_URL, DEFAULT_ANSWER_TIME_MS
from src.utils.headers import merge_headers
from src.utils.types import CaptchaData, CommentPayload


class CommentApi:
    def __init__(self, session: requests.Session) -> None:
        self._session = session

    def get_captcha(self) -> requests.Response:
        return self._session.get(f"{API_URL}/api/comments/captcha", timeout=10)

    def get_comments(self, photo_id: int, page: int = 0, size: int = 5) -> requests.Response:
        return self._session.get(
            f"{API_URL}/api/comments",
            params={"photoId": photo_id, "page": page, "size": size},
            timeout=10,
        )

    def create(
        self,
        payload: CommentPayload,
        captcha: CaptchaData,
        custom_headers: dict[str, str] | None = None,
    ) -> requests.Response:
        answer_time_ms = captcha.get("answerTimeMs", DEFAULT_ANSWER_TIME_MS)
        headers = merge_headers(
            {
                "X-Captcha-Session-Id": captcha["sessionId"],
                "X-Captcha-Answer": str(captcha["answer"]),
                "X-Answer-Time-Ms": str(answer_time_ms),
                **(custom_headers or {}),
            }
        )
        return self._session.post(
            f"{API_URL}/api/comments",
            json=dict(payload),
            headers=headers,
            timeout=10,
        )
